src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -921,45 +921,55 @@ public static bool IsValidParams(Symbol member)
             return final.IsParams && ((ParameterSymbol)final.OriginalDefinition).Type.IsSZArray();
         }
 
-        private static bool IsOverride(Symbol overridden, Symbol overrider)
+        /// <summary>
+        /// Does <paramref name="moreDerivedOverride"/> override <paramref name="member"/> or the
+        /// thing that it originally overrides, but in a more derived class?
+        /// </summary>
+        private static bool IsMoreDerivedOverride(Symbol member, Symbol moreDerivedOverride)
         {
-            if (TypeSymbol.Equals(overridden.ContainingType, overrider.ContainingType, TypeCompareKind.ConsiderEverything2) ||
-                !MemberSignatureComparer.SloppyOverrideComparer.Equals(overridden, overrider))
+            if (!moreDerivedOverride.IsOverride ||
+                !IsBaseClass(derivedType: moreDerivedOverride.ContainingType, baseType: member.ContainingType) ||
+                !MemberSignatureComparer.SloppyOverrideComparer.Equals(member, moreDerivedOverride))
             {
                 // Easy out.
                 return false;
             }
 
-            // Does overrider override overridden?
-            var current = overrider;
-            while (true)
-            {
-                if (!current.IsOverride)
-                {
-                    return false;
-                }
-                current = current.GetOverriddenMember();
+            // Rather than following the member.GetOverriddenMember() chain, we check to see if both
+            // methods ultimately override the same original method.  This addresses issues in binary compat
+            // scenarios where the override chain may skip some steps.
+            // See https://github.com/dotnet/roslyn/issues/45798 for an example.
+            return moreDerivedOverride.GetLeastOverriddenMember(accessingTypeOpt: null).OriginalDefinition ==
```

Comment:
>OriginalDefinition [](start = 88, length = 18)

Comparing original definitions feels fragile. Methods in generic types that are constructed with different type arguments cannot override the same method, even when the overridden methods correspond to the same original definition.

Reply:
We've already verified that the methods' containing types are in a precise inheritance relationship.

---
In reply to: [461770335](https://github.com/dotnet/roslyn/pull/46367#discussion_r461770335) [](ancestors = 461770335)

Reply:
> We've already verified that the methods' containing types are in a precise inheritance relationship.

Where and how, is that the other place I made a comment about the TypeCompareKind used?

---
In reply to: [461838954](https://github.com/dotnet/roslyn/pull/46367#discussion_r461838954) [](ancestors = 461838954,461770335)

Reply:
In general, I do not believe taking shortcuts like that buy us anything, they do not simplify code, do not make it easer to understand and maintain. I believe that it would be more robust if we don't take them without a real benefit that can be demonstrated/measured.

---
In reply to: [461938985](https://github.com/dotnet/roslyn/pull/46367#discussion_r461938985) [](ancestors = 461938985,461838954,461770335)

Reply:
> Where and how, is that the other place I made a comment about the TypeCompareKind used?

The where is on line 931.  This preserves the behavior of this code for existing scenarios not involving covariant returns.  I do not intend to modify or relax the behavior of this code for existing scenarios, so this code is as intended.

---
In reply to: [461948173](https://github.com/dotnet/roslyn/pull/46367#discussion_r461948173) [](ancestors = 461948173,461938985,461838954,461770335)